app/markdown-it-plugins/cite.js

```diff
@@ -0,0 +1,177 @@
+  md.renderer.rules.cite_ref = (tokens, idx) => { // also take: options, env, self
+    const token = tokens[idx];
+    const key = token.meta.key;
+
+    if (!token.meta.htmlKey || null === token.meta.htmlKey) {
+      return `<span title="This reference is invalid" class="invalid-ref">[@${key}]</span>`;
+    }
+
+    return token.meta.htmlKey;
+  };
+
+  // override fence block parser
+  md.block.ruler.at('fence', (state, startLine, endLine, silent) => {
+    let len;
+    let nextLine;
+    let mem;
+    let haveEndMarker = false;
+    let pos = state.bMarks[startLine] + state.tShift[startLine];
+    let max = state.eMarks[startLine];
+
+    if (pos + 3 > max) { return false; }
+
+    const marker = state.src.charCodeAt(pos);
+
+    if (marker !== 0x7E/* ~ */ && marker !== 0x60 /* ` */) {
+      return false;
+    }
+
+    // scan marker length
+    mem = pos;
+    pos = state.skipChars(pos, marker);
+
+    len = pos - mem;
+
+    if (len < 3) { return false; }
+
+    const markup = state.src.slice(mem, pos);
+    const params = state.src.slice(pos, max);
+
+    if (params.indexOf('`') >= 0) { return false; }
+
+    // Since start is found, we can report success here in validation mode
+    if (silent) { return true; }
+
+    // search end of block
+    nextLine = startLine;
+
+    for (;;) {
+      nextLine++;
+      if (nextLine >= endLine) {
+        // unclosed block should be autoclosed by end of document.
+        // also block seems to be autoclosed by end of parent
+        break;
+      }
+
+      pos = mem = state.bMarks[nextLine] + state.tShift[nextLine];
+      max = state.eMarks[nextLine];
+
+      if (pos < max && state.sCount[nextLine] < state.blkIndent) {
+        // non-empty line with negative indent should stop the list:
+        // - ```
+        //  test
+        break;
+      }
+
+      if (state.src.charCodeAt(pos) !== marker) { continue; }
+
+      if (state.sCount[nextLine] - state.blkIndent >= 4) {
+        // closing fence should be indented less than 4 spaces
+        continue;
+      }
+
+      pos = state.skipChars(pos, marker);
+
+      // closing code fence must be at least as long as the opening one
+      if (pos - mem < len) { continue; }
+
+      // make sure tail has spaces only
+      pos = state.skipSpaces(pos);
+
+      if (pos < max) { continue; }
+
+      haveEndMarker = true;
+      // found!
+      break;
+    }
+
+    // If a fence has heading spaces, they should be removed from its inner block
+    len = state.sCount[startLine];
+
+    state.line = nextLine + (haveEndMarker ? 1 : 0);
+
+    const token = state.push('fence', 'code', 0);
+    token.info = params;
+    token.content = state.getLines(startLine + 1, nextLine, len, true);
+    token.markup = markup;
+    token.map = [startLine, state.line];
+
+    if (!state.env.citations) {
+      state.env.citations = [];
+    }
+
+    if ('cite' === token.info) {
+      state.env.citations = state.env.citations.concat(
```

Comment:
Note for myself: there I have to build this `citations` array so that I can wire references in the text with citations in bibtex blocks.